ADBDEV-5465: Split temp tables storage from temp files storage  (#889)

silent-observer opened the pull request:
Split temp tables storage from temp files storage  (#889)

Some operations create temporary files on disk. The growth of temporary files in
conjunction with temporary tables can lead to 100% usage of the data directory
and an emergency stop of the DBMS.

Implement temp_spill_files_tablespaces GUC that controls tablespaces where
temporary files are stored. Temporary tables will still be stored in
temp_tablespaces.

In case temp_spill_files_tablespaces is not set: the behavior will remain the
same, and both files and tables will be stored in the specified table spaces
according to the list in temp_tablespaces, or if it's also empty, in system
table spaces.

If the temp_spill_files_tablespaces is not empty, but temp_tablespaces is empty,
only temporary files will be saved in the specified table spaces. Temporary
table files will still be stored in the system tablespace.

If temp_spill_files_tablespaces is set to the empty string ("") or 'pg_default',
it will not fall back to temp_tablespaces and will instead use the default
tablespace.

Changes from original commit:
1. Remove DITA format docs since they are missing in GPDB 7.
2. OpenNamedTemporaryFile is missing in GPDB 7, no fix needed.
3. Use GPDB 7 implementation of get_session_temp_tablespace instead of GPDB 6.
4. Additionally fix GetTempTablespaces (used by SharedFileSetInit) which now
   uses spill file tablespaces if there are any.
5. There is no GetSessionTempTableSpace, so call get_session_temp_tablespace
   directly from GetNextTempTableSpace.

Test-related changes:
1. Add file_monitor.c from GPDB 6.
2. Add GNUMakefile entry for file_monitor from GPDB 6.
3. Add gp_tablespace_tmppath function to regress.c from GPDB 6.
4. Add plpython functions from GPDB 6 with fixes for GPDB 7 (switch to
   plpython3, add SETOF, add utf-8 encoding).
5. Fix last query from gp_tablespace_file_report to execute on primary
   segments using gp_segment_configuration.

(cherry picked from commit 4bb482a12db5132bf94911b7c96b4ea0d2df3224)

---

Fix temp_spill_files_tablespaces GUC sync (#1074)

Previously GUC sync worked by sending SET commands to segments. However,
some values (for example the empty string for quoted GUCs) cannot be set this
way. This affects specifically temp_spill_files_tablespaces since "" and
"\\"\\"" have different semantic meaning for it.

This patch changes the way GUC sync works by using pg_catalog.set_config()
function instead of SET commands. This function sets the value of the GUC
directly without any quoting issues, and so now empty strings are handled
correctly.

(cherry picked from commit 993b6c4b24f3773762fe924851da9f6518cb002d)

---

Note: do not squash the commit to preserve authorship.

bandetto committed:
Split temp tables storage from temp files storage  (#889)

Some operations create temporary files on disk. The growth of temporary files in
conjunction with temporary tables can lead to 100% usage of the data directory
and an emergency stop of the DBMS.

Implement temp_spill_files_tablespaces GUC that controls tablespaces where
temporary files are stored. Temporary tables will still be stored in
temp_tablespaces.

In case temp_spill_files_tablespaces is not set: the behavior will remain the
same, and both files and tables will be stored in the specified table spaces
according to the list in temp_tablespaces, or if it's also empty, in system
table spaces.

If the temp_spill_files_tablespaces is not empty, but temp_tablespaces is empty,
only temporary files will be saved in the specified table spaces. Temporary
table files will still be stored in the system tablespace.

If temp_spill_files_tablespaces is set to the empty string ("") or 'pg_default',
it will not fall back to temp_tablespaces and will instead use the default
tablespace.

Changes from original commit:
1. Remove DITA format docs since they are missing in GPDB 7.
2. OpenNamedTemporaryFile is missing in GPDB 7, no fix needed.
3. Use GPDB 7 implementation of get_session_temp_tablespace instead of GPDB 6.
4. Additionally fix GetTempTablespaces (used by SharedFileSetInit) which now
   uses spill file tablespaces if there are any.
5. There is no GetSessionTempTableSpace, so call get_session_temp_tablespace
   directly from GetNextTempTableSpace.

Test-related changes:
1. Add file_monitor.c from GPDB 6.
2. Add GNUMakefile entry for file_monitor from GPDB 6.
3. Add gp_tablespace_tmppath function to regress.c from GPDB 6.
4. Add plpython functions from GPDB 6 with fixes for GPDB 7 (switch to
   plpython3, add SETOF, add utf-8 encoding).
5. Fix last query from gp_tablespace_file_report to execute on primary
   segments using gp_segment_configuration.

(cherry picked from commit 4bb482a12db5132bf94911b7c96b4ea0d2df3224)

bandetto commented on src/test/regress/input/temp_tablespaces.source:
```diff
@@ -138,3 +189,249 @@ drop tablespace mytempsp1;
+
+--
+-- Global GUC value tests:
+-- FIXME: We shouldn't be forced to use \c to update GUCs on segments.
```
We fixed this issue earlier. Now `gpstop -u` (in ignore blocks below) shouldn't require running `\c` to reload the GUCs.

silent-observer replied:
What has to be changed here, just removing the commend and `\c`, or is there something else?

bandetto replied:
Just remove `\c` in blocks below. The GUCs should be updated correctly now. `\c` causes a reconnect and was necessary before fixing the `SIGHUP` bug.

bandetto commented on src/test/regress/input/temp_tablespaces.source:
```diff
@@ -37,6 +37,57 @@ create tablespace mytempsp2 location '@testtablespace@_mytempsp2';
+if rc != 0:
+    output, error = process.communicate()
+    raise Exception("error=%s" % str(error))
+return []
```
Can you do this and change the return type to just `void` Same with `..watch_stop()`?
```suggestion
return
```

silent-observer replied:
GPDB 7 outputs
```
ERROR:  EXECUTE ON ALL SEGMENTS is only supported for set-returning functions
```
Also, all `EXECUTE ON ALL SEGMENTS` in GPDB 7 tests seem to use `SETOF`, even for `void`. Are there other solutions for this?

bandetto replied:
I looked into it previously, and I guess not. For GPDB 7, there's now an explicit check that prohibits the creation of non-set-returning functions with EXECUTE ON clause.

bandetto commented on src/test/regress/output/temp_tablespaces.source:
```diff
@@ -227,3 +270,300 @@ drop tablespace mytempsp1;
+ default       | default
+(3 rows)
+
+-- Both files and the table should be in temp_tables.
+SET temp_tablespaces = 'temp_tables';
+RESET temp_spill_files_tablespaces;
+SELECT * FROM gp_tablespace_file_report(1);
+ temp_table_in | temp_files_in 
+---------------+---------------
+ temp_tables   | temp_tables
+ temp_tables   | temp_tables
+ temp_tables   | temp_tables
+(3 rows)
```
The tests failed here.

silent-observer replied:
This seems to be another GUC-related issue. At some point I've managed to achieve
```
\! gpconfig -c temp_tablespaces -v 'temp_tables'
\! gpstop -u
SHOW temp_tablespaces;
 temp_tablespaces 
------------------
 temp_tables
(1 row)

\! gpconfig -r temp_tablespaces
\! gpstop -u
SHOW temp_tablespaces;
 temp_tablespaces 
------------------
 temp_tables
(1 row)
```
while testing locally, although I can't replicate this reliably yet. I'll research this more.

Note: leaving the last `\c` makes the test pass.

silent-observer replied:
Yup, GUC sync issue. Reproduction scenario (both on GPDB 6 and GPDB 7):
```
SET temp_tablespaces = 'pg_default';
SHOW temp_tablespaces;
\! gpconfig -r temp_tablespaces
\! gpstop -u
SHOW temp_tablespaces;
```
outputs `pg_default` on both SHOWs.

silent-observer replied:
This is indeed a GUC sync issue. After GUC reset, coordinator sends `SET temp_spill_files_tablespaces TO ''` to segments, except in case of this specific GUC `RESET` is not equivalent to setting the default value (empty string) explicitly. In general this behavior seems very confusing.

silent-observer replied:
Fixed this issue for GPDB 6, cherry picked the commit here.

silent-observer committed:
Fix temp_spill_files_tablespaces GUC sync (#1074)

Previously GUC sync worked by sending SET commands to segments. However,
some values (for example the empty string for quoted GUCs) cannot be set this
way. This affects specifically temp_spill_files_tablespaces since "" and
"\"\"" have different semantic meaning for it.

This patch changes the way GUC sync works by using pg_catalog.set_config()
function instead of SET commands. This function sets the value of the GUC
directly without any quoting issues, and so now empty strings are handled
correctly.

Ticket: ADBDEV-6438
(cherry picked from commit 993b6c4b24f3773762fe924851da9f6518cb002d)

Stolb27 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5465
